### ISSUE TITLE
Deserialization of compressed data is sluggish

### ISSUE THREAD

**Issue by crusaderky** [dask/distributed#7433]
If lz4 is installed and a chunk can be compressed efficiently, it is compressed automatically.
This happens both in network comms (configurable) and upon spilling (not configurable).

While working on https://github.com/coiled/coiled-runtime/pull/629, I found out that deserializing a compressed frame is abysmally slow:

![image](https://user-images.githubusercontent.com/6213168/209113228-9ca8f43e-1944-4c9b-965d-8c47b884dc2a.png)

When deserialization is triggered by network comms, it is offloaded to a separate thread for any message larger than 10 MiB (`distributed.comm.offload`). Note that a message will contain multiple small keys if possible.
I did not test if the GIL is released while this happens.

When deserialization is triggered by unspilling, the whole event loop is temporarily blocked.

### Microbenchmark
Serializing a 128 MiB compressible numpy array takes 10.4ms; there's a modest overhead on top of the raw call to lz4, which takes 9.73ms.
Deserializing the same takes a whopping 144ms. The raw lz4 decompress  takes 35 *micro*seconds.

```python
import io
import lz4
import numpy
from distributed.protocol import deserialize_bytes, serialize_bytelist

a1 = numpy.random.random((4096, 4096))  # 128 MiB, uncompressible
a2 = numpy.ones((4096, 4096))  # 128 MiB, highly compressible

buf1 = io.BytesIO()
buf2 = io.BytesIO()
buf1.writelines(serialize_bytelist(a1))
buf2.writelines(serialize_bytelist(a2))
b1 = buf1.getvalue()
b2 = buf2.getvalue()
b3 = lz4.frame.compress(a2)

print("uncompressible size", len(b1))
print("compressible size", len(b2))
print("raw lz4 size", len(b3))
print()

print("serialize uncompressible")
%timeit serialize_bytelist(a1)
print("serialize compressible")
%timeit serialize_bytelist(a2)
print("raw lz4 compress")
%timeit lz4.frame.compress(a2)
print()

print("deserialize uncompressible")
%timeit deserialize_bytes(b1)
print("deserialize compressible")
%timeit deserialize_bytes(b2)
print("raw lz4 decompress")
%timeit lz4.frame.compress(b3)
```
Output:
```
uncompressible size 134217969
compressible size 526635
raw lz4 size 566291

serialize uncompressible
101 µs ± 13.8 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
serialize compressible
10.4 ms ± 516 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
raw lz4 compress
9.73 ms ± 98.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

deserialize uncompressible
19 ms ± 92.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
deserialize compressible
144 ms ± 770 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
raw lz4 decompress
35.4 µs ± 748 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

`deserialize_bytes` `serialize_bytelist` in the code below is exactly what `distributed.spill` uses. However, I measured the timings in `distributed.comm.tcp` and I got the same number (134ms).

**Comment by crusaderky** [comment]
Notably, a lot of the tests in coiled-runtime start from `da.random.random()`, which produces uncompressible data, and will not show this issue.

**Comment by shughes-uk** [comment]
Replicated on 

Python: 3.11.0
macOS: 13.1
arch : arm64
lz4 : 4.0.2
distributed : 2022.12.1
numpy : 1.24.0
```
uncompressible size 134217969
compressible size 526635
raw lz4 size 566291

serialize uncompressible
39.5 µs ± 3.99 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
serialize compressible
8.35 ms ± 48.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
raw lz4 compress
7.78 ms ± 49.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

deserialize uncompressible
11.3 ms ± 577 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
deserialize compressible
70.4 ms ± 659 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
raw lz4 decompress
40.6 ms ± 479 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by crusaderky** [comment]
TL,DR; It's death by deep-copy. We're spending 89ms doing actual decompression and 45ms doing deep-copies.


I think we're spendingat least part of the problem is death by deep-copy.

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.134    0.134 distributed/protocol/serialize.py:662(deserialize_bytes)
        1    0.000    0.000    0.098    0.098 distributed/protocol/compression.py:193(decompress)
        1    0.000    0.000    0.098    0.098 distributed/protocol/compression.py:195(<listcomp>)
        2    0.098    0.049    0.098    0.049 {built-in method lz4.block._block.decompress}
        1    0.000    0.000    0.036    0.036 distributed/protocol/serialize.py:473(merge_and_deserialize)
        1    0.036    0.036    0.036    0.036 {method 'join' of 'bytearray' objects}
```
The 128 MiB input buffer is split in two frames (`distributed.block.shard: 64 MiB`).
The two calls to `lz4.block._block.decompress`:
1. write the output to a local raw buffer `dest`:
https://github.com/python-lz4/python-lz4/blob/79370987909663d4e6ef743762768ebf970a2383/lz4/block/_block.c#L349-L359
3. then, they deep-copy to the output bytes object:
https://github.com/python-lz4/python-lz4/blob/79370987909663d4e6ef743762768ebf970a2383/lz4/block/_block.c#L389
4. The two bytes objects are not contiguous. Hence, they are deep-copied again into their final destination by `merge_and_deserialize`

Fixing this will require an upstream PR that lets us pass an `out=` parameter.

It can't be the whole story though.
A deep-copy of 128 MiB takes 9ms on my computer:
```python
a2copy = numpy.empty_like(a2)
%timeit a2copy[:] = a2
9.15 ms ± 46.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
So we've justified at 18ms out of 134ms.

If I increase `distributed.block.shard` to 128 MiB, the profile changes - but not in a very good way:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.113    0.113 distributed/protocol/serialize.py:662(deserialize_bytes)
        1    0.000    0.000    0.094    0.094 distributed/protocol/compression.py:193(decompress)
        1    0.000    0.000    0.094    0.094 distributed/protocol/compression.py:195(<listcomp>)
        1    0.094    0.094    0.094    0.094 {built-in method lz4.block._block.decompress}
        1    0.000    0.000    0.019    0.019 distributed/protocol/serialize.py:473(merge_and_deserialize)
        1    0.000    0.000    0.019    0.019 distributed/protocol/serialize.py:371(deserialize)
        1    0.000    0.000    0.019    0.019 distributed/protocol/serialize.py:56(dask_loads)
        1    0.000    0.000    0.019    0.019 distributed/utils.py:746(wrapper)
        1    0.000    0.000    0.019    0.019 distributed/protocol/numpy.py:114(deserialize_numpy_ndarray)
        1    0.000    0.000    0.019    0.019 numpy/core/_asarray.py:22(require)
        1    0.019    0.019    0.019    0.019 {method 'copy' of 'numpy.ndarray' objects}
```
We can see that numpy is performing a deep copy - which should not happen. So even here we should be able to shave off 19ms off in numpy deserialization and 9ms in PyBytes_FromStringAndSize.

**Comment by crusaderky** [comment]
Worth noting that blosc (removed in #5269) does offer deserialization to already existing memory (albeit unpythonic and unsafe).

**Comment by crusaderky** [comment]
This works, thanks to #5208:

```python
import dask
import distributed
import numpy

c = distributed.Client(n_workers=2, memory_limit="8 GiB")
x = c.submit(numpy.random.random, 4 * 2**30 // 8, workers=[0])
y = c.submit(lambda x: None, x, workers=[1])
```

Replace `numpy.random.random` with `numpy.ones`, and the receiving worker will be killed by the nanny.

**Comment by crusaderky** [comment]
> TL,DR; It's death by deep-copy. We're spending 85ms doing actual decompression and 49ms doing deep-copies.

Another important note:
The actual decompression part releases the GIL. So, as long as serialization/deserialization iss offloaded to another thread (#4424), it's fine.
To my understanding, the deep copies all hold the GIL while they are performed.

**Comment by mrocklin** [comment]
I would be curious to see if turning off compression entirely affects the benchmark suite.  That might give us a better holistic understanding of how much impact this has.  Is that easy to test?

**Comment by crusaderky** [comment]
> I would be curious to see if turning off compression entirely affects the benchmark suite. That might give us a better holistic understanding of how much impact this has. Is that easy to test?

Yes but a lot of data in the benchmark suite is created with numpy.random.random - which is uncompressible - so it's not necessarily (I'd dare say, unlikely) representative of real world data.

**Comment by crusaderky** [comment]
I think the whole thing can be dropped from 143ms to 18ms. I think we were too hasty to drop support for blosc - the performance benefit is huge.

```python
import pickle
from time import time
import blosc


def compress(frame):
    if isinstance(frame, pickle.PickleBuffer):
        mv = memoryview(frame)
        return blosc.compress(mv, clevel=5, cname="blosclz", typesize=mv.itemsize)
    else:
        return blosc.compress(frame, clevel=5, cname="blosclz")

a = numpy.linspace(0, 1, 128 * 2**20 // 8)
# a = numpy.ones(128 * 2**20 // 8)

t0 = time()
frames = []
pik = pickle.dumps(a, protocol=5, buffer_callback=frames.append)
t1 = time()
frames.insert(0, pik)
print([len(memoryview(frame)) for frame in frames])
cframes = [compress(frame) for frame in frames]
print([len(frame) for frame in cframes])
t2 = time()
frames2 = [blosc.decompress(frame, as_bytearray=True) for frame in cframes]
t3 = time()
a2 = pickle.loads(frames2[0], buffers=frames2[1:])
t4 = time()
numpy.testing.assert_array_equal(a, a2)

print("dumps", t1 - t0)
print("compress", t2 - t1)
print("decompress", t3 - t2)
print("loads", t4 - t3)
```
```

[120, 16777216]
[136, 1069658]
dumps 0.0005013942718505859
compress 0.006549835205078125
decompress 0.012061119079589844
loads 0.0056269168853759766
```

On a separate topic - what's the purpose of sharding? I can't see how it makes us avoid any memory spikes anyway, as we always have all the sharded frames in memory before we do anything with them?

**Comment by mrocklin** [comment]
While it's true that we won't be able to see benefits of compressor we will
be able to see the costs pretty clearly.  If we turn off compression and
nothing changes then we gain the option of ignoring this for a while.

Sharding is important as data sizes grow to 2GB, when many algorithms break
down.  Below that we're guessing that it's nice (smoother maybe).

**Comment by crusaderky** [comment]
> While it's true that we won't be able to see benefits of compressor we will be able to see the costs pretty clearly.

Actually, we won't. `maybe_compress` tries compressing 10 kiB and, if that compresses to more than 90%, it doesn't attempt to compress the rest. This exercise is quite negligible, CPU wise.

> Sharding is important as data sizes grow to 2GB, when many algorithms break down. 

Are you talking about the network stack?
Do you have any specific examples in mind? I'd like to discuss this more in depth (at our next stand-up?)
If it's a networking problem, we could move sharding downstream of compression - that alone would be greatly beneficial.

**Comment by mrocklin** [comment]
> Actually, we won't. maybe_compress tries compressing 10 kiB and, if that compresses to more than 90%, it doesn't attempt to compress the rest. This exercise is quite negligible, CPU wise.

Ah, good point

> Do you have any specific examples in mind? 

```python
In [1]: import lz4.block

In [2]: lz4.block.compress(b'0' * int(3e9))
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In [2], line 1
----> 1 lz4.block.compress(b'0' * int(3e9))

OverflowError: Input too large for LZ4 API
```

**Comment by crusaderky** [comment]
Then I think we should seriously reconsider blosc. Besides being 5x faster, it's also the only API that allows decompressing to an already existing bytearray. With sharding upstream of compression, that's the only way to avoid deep copies.

**Comment by crusaderky** [comment]
To recap the discussion: both the time spent and the memory flares are caused by the conjunction of the following:

- The python lz4 library performs an unnecessary deep copy before it returns the array. Needs upstream PR.
- decompression when unspilling is not offloaded out of the main thread, unlike decompression when receiving from the network
(#4424)
- distributed used to perform a deep copy when unpickling a writeable numpy array from a read-only bytes buffer. Closed by #7437 for single-shard compressed data and by #7435 for spilled uncompressed data.
- lz4 is very slow to decompress, compared to blosc
- sharded buffers are decompressed to separate, non-contiguous buffers which are necessarily deep-copied afterwards. To fix this, the compression library needs to support decompressing to an existing buffer. blosc already supports it; we could contribute a patch to lz4 to add the feature.

Additionally,
- coiled_runtime CI for the most part uses uncompressible data, which does not show the problem.

**Comment by mrocklin** [comment]
> Then I think we should seriously reconsider blosc

I don't strongly object.  If we add blosc again I think that we should try to add it in a normal way similar to other compressors, rather than the way that it was implemented before, as part of the serializer.  This caused frequent complications.

> lz4 is very slow to decompress, compared to blosc

My recollection is that it's faster than 1GB/s, and so probably very cheap relative to other things in the pipeline.  "Slower than blosc" doesn't seem motivating to me.  I think that the thing to beat here is "a common bottleneck in common workloads".

> coiled_runtime CI for the most part uses uncompressible data, which does not show the problem

I agree that having benchmark results would be helpful in motivating the prioritization of this work.  Right now the original motivation (dot-product with spill on compressible data) feels a little niche to me given the amount of effort.  Of course, if you're having fun then don't let me stop you 🙂

**Comment by mrocklin** [comment]
Of course, maybe this actually applies in lots of common cases (every time we move compressible data bigger than a few megabytes and every time we spill) and has been having a large but silent impact.  If so, hooray 🎉.  In general, +1 to benchmarking.  I would hope that this would come up in some of the realistic benchmarks?  Vorticitiy?  Dataframe joins?

**Comment by crusaderky** [comment]
>> lz4 is very slow to decompress, compared to blosc

> My recollection is that it's faster than 1GB/s

680 MiB/s compression
1330 MiB/s decompression (<= 64 MiB buffers)
920 MiB/s decompression (> 64 MiB buffers)

Which is, in fact, extremely noticeable - discussion and code on #7655.
